src/users/userSelectors.js

```diff
@@ -169,3 +179,122 @@ const getActiveUsersById: Selector<Map<UserId, UserOrBot>> = createSelector(
+      // The web client appears to treat these two cases identically.
+      // So we do the same.
+      return { displayType: 'text', text: value };
+
+    case 3: {
+      // CustomProfileFieldType.Choice
+      // TODO(server): This isn't really documented.  But see chat thread:
+      //   https://chat.zulip.org/#narrow/stream/378-api-design/topic/custom.20profile.20fields/near/1383005
+      const choices = JSON.parse(realmField.field_data);
+      return { displayType: 'text', text: choices[value].text };
+    }
+
+    case 4: {
+      // CustomProfileFieldType.Date
+      // TODO(server): The value's format is undocumented, but empirically
+      //   it's a date in ISO format, like 2000-01-01.
+      // That's readable as is, but:
+      // TODO format this date using user's locale.
+      return { displayType: 'text', text: value };
+    }
+
+    case 5: // CustomProfileFieldType.Link
+      // This isn't real clearly documented; but the `value` is just the URL.
+      return { displayType: 'link', text: value, url: new URL(value) };
+
+    case 7: {
+      // CustomProfileFieldType.ExternalAccount
+      // TODO(server): This is undocumented.  See chat thread:
+      //   https://chat.zulip.org/#narrow/stream/378-api-design/topic/external.20account.20custom.20profile.20fields/near/1387213
+      const realmData: { subtype: string, url_pattern?: string } = JSON.parse(
+        realmField.field_data,
+      );
+      const { subtype, url_pattern } = realmData;
+      const pattern = url_pattern ?? realmDefaultExternalAccounts.get(subtype)?.url_pattern;
+      const url = pattern == null ? undefined : new URL(pattern.replace('%(username)s', value));
+      if (!url) {
+        logging.warn(
+          `Missing url_pattern for custom profile field of type ExternalAccount, subtype '${subtype}'`,
+        );
+      }
+      return { displayType: 'link', text: value, url };
+    }
+
+    case 6: {
+      // CustomProfileFieldType.User
+      // TODO(server): This is completely undocumented.  The key to
+      //   reverse-engineering it was:
+      //   https://github.com/zulip/zulip/blob/18230fcd9/static/js/settings_account.js#L247
+      const userIds: UserId[] = JSON.parse(value);
+      return { displayType: 'users', userIds };
+    }
+
+    default:
+      ensureUnreachable(type);
+      logging.warn(`Invalid custom profile field type: ${type}`);
+      return undefined;
+  }
+}
+
+/**
+ * The given user's values for custom profile fields, interpreted to make usable.
+ *
+ * These are in the order a client is expected to present them in.  Any
+ * fields that are configured on the realm but that the user has no chosen
+ * value for are omitted.
+ *
+ * This selector does no caching.  Meant for use inside `React.useMemo`.
+ */
+export function getCustomProfileFieldsForUser(
+  realm: RealmState,
+  user: UserOrBot,
+): Array<{| +fieldId: number, +name: string, +value: CustomProfileFieldValue |}> {
```

Comment:
This is the function's return type, so it's purely in an output/covariant position -- which means that for the caller `Array` means more flexibility without losing any flexibility.

Sometimes that flexibility comes in handy, e.g. for wanting to sort the array. In `recipient.js` in particular there've been some places where one function was returning a `$ReadOnlyArray` and the caller wanted to sort it, and it required some refactoring.

Here this function is constructing the array fresh (rather than pulling it verbatim from some existing data structure), so there's no constraint requiring it to be read-only -- so we might as well leave the caller the flexibility to use that fact.